node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1601,81 +1901,137 @@ fn get_era_id_for_validators_retrieval(
-        .get_available_block_range_from_storage()
-        .await;
-
-    let mut chain_sync_context = ChainSyncContext::new(
-        &effect_builder,
-        &config,
-        &metrics,
-        locally_available_block_range_on_start,
-    );
-
-    let trusted_block_header =
-        fetch_and_store_initial_trusted_block_header(&chain_sync_context, &metrics, trusted_hash)
-            .await?;
-
-    chain_sync_context.set_trusted_block_header(&trusted_block_header);
-
-    verify_trusted_block_header(&chain_sync_context)?;
+    let ctx = match ChainSyncContext::new(&effect_builder, &config, &metrics).await? {
+        Some(ctx) => ctx,
+        None => return Ok(FastSyncOutcome::ShouldCommitGenesis),
+    };
+    verify_trusted_block_header(&ctx)?;
+
+    // We should have at least one block header in storage now as a result of calling
+    // `ChainSyncContext::new`.
+    let mut highest_block_header =
+        match effect_builder.get_highest_block_header_from_storage().await {
+            Some(block_header) => block_header,
+            None => return Err(Error::NoHighestBlockHeader),
+        };
 
-    if handle_emergency_restart(&chain_sync_context).await? {
-        return Ok(*trusted_block_header);
+    if let Some(outcome) = should_emergency_upgrade(&ctx, &highest_block_header).await? {
+        return Ok(outcome);
     }
 
-    if handle_upgrade(&chain_sync_context).await? {
-        return Ok(*trusted_block_header);
+    if let Some(outcome) = should_upgrade(&ctx, &highest_block_header).await? {
+        return Ok(outcome);
     }
 
-    let (trusted_key_block_info, most_recent_block_header) = if config.sync_to_genesis() {
-        sync_to_genesis(&chain_sync_context).await?
-    } else {
-        fast_sync(&chain_sync_context).await?
-    };
+    let trusted_key_block_info = get_trusted_key_block_info(&ctx).await?;
+    let most_recent_block_header = fast_sync(&ctx, &trusted_key_block_info).await?;
 
     // Iterate forwards, fetching each full block and deploys but executing each block to generate
     // global state. Stop once we get to a block in the current era.
-    let most_recent_block_header = execute_blocks(
-        &most_recent_block_header,
-        &trusted_key_block_info,
-        &chain_sync_context,
-    )
-    .await?;
+    let most_recent_block_header =
+        execute_blocks(&most_recent_block_header, &trusted_key_block_info, &ctx).await?;
+
+    // If we just committed an emergency upgrade and are re-syncing right after this, potentially
+    // the call to `fast_sync` and `execute_blocks` could yield a `most_recent_block_header` which
+    // is older than the `highest_block_header` (which is the immediate switch block of the
+    // upgrade).  Ensure we take the actual highest block to allow consensus to be initialised
+    // properly.
+    if most_recent_block_header.height() > highest_block_header.height() {
+        highest_block_header = most_recent_block_header;
+    }
```

Comment:
Won't it _always_ be `most_recent_block_header` anyway? Could `highest_block_header` ever be higher than that?

If not it might be safest to `drop(highest_block_header)` after the `should_upgrade` call, and just use `most_recent_block_header` here?

Reply:
OK, thanks for clarifying! :+1: 
(I would have thought that in this second run of fast sync, they'd load the immediate switch block from disk and make that their `most_recent_block_header`.)